### PR TITLE
Refactor code for index directory

### LINKED ISSUE #818: Complex validation for index files
We currently only validate whether the version of moPepGen, python, and biopython match with the index. However, commands that use cleavage params should also be validated for those. It causes problems if you want callVariant to use arg-c but the index is trypsin.

### PR DESCRIPTION
**Pull request by zhuchcn**
## Description

I don't like the way we handle the index directory and files currently. Right now we store versions of moPepGen, python and biopython in genome.pkl, annotation_gene.idx, annotation_tx.idx, and proteome.pkl. So here I created a metadata.json file to store all essential parameters, including versions, cleavage parameters (enzyme, miscleavages, etc), and genome/annotation source (GENCODE or ENSEMBL).

I also created an `IndexDir` class to handle the index directory responsible for loading and dumping different index files, making the code more organized.

Closes #818  

## Checklist

- [X] This PR does NOT contain [PHI](https://ohrpp.research.ucla.edu/hipaa/) or germline genetic data.  A repo may need to be deleted if such data is uploaded. Disclosing PHI is a [major problem](https://healthitsecurity.com/news/ucla-health-reaches-7.5m-settlement-over-2015-breach-of-4.5m).
- [X] This PR does NOT contain molecular files, compressed files, output files such as images (*e.g.* `.png`, .`jpeg`), `.pdf`, `.RData`, `.xlsx`, `.doc`, `.ppt`, or other non-plain-text files.  To automatically exclude such files using a [.gitignore](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files) file, see [here](https://github.com/uclahs-cds/template-base/blob/main/.gitignore) for example.
- [X] I have read the [code review guidelines](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3187646/Code+Review+Guidelines) and the [code review best practice on GitHub check-list](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List).
- [X] The name of the branch is meaningful and well formatted following the [standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List), using [AD_username (or 5 letters of AD if AD is too long)]-[brief_description_of_branch].
- [X] I have added the major changes included in this pull request to the `CHANGELOG.md` under the next release version or unreleased, and updated the date.
- [X] All test cases passed locally.

### CODE REVIEW

**Review by lydiayliu — APPROVED** [review]
I've always wondered, what happens if you use an index where the cleavage parameters are different from that in `callVariant`? Where are the cleavage paramters used in generateIndex (I assume just the canonical proteome fasta)? Will it be reprocessed on the fly if things do differ or would the results be erroneous? (For example when indexing I did peptide length 7-25, no miscleavages, for `callVariant` I did peptide 4-35, 2 miscleavages

Also this is a pretty large change, we should run some real data for sanity check

File: moPepGen/index.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        if file.suffix.lower() == '.gz':
+            if symlink:
+                symlink = False
+                logger(
+                    "--gtf-symlink was suppressed because compressed GTF file was received. "
+                )
+        elif file.suffix.lower() != '.gtf':
+            raise ValueError(f"Cannot handle gtf file {file}")
+
+
```
[the remark this answers is not in the document]
Comment (zhuchcn) [reply moPepGen/index.py:79]: I'll fix this in the next PR

**Comment by zhuchcn** [comment]
The only place where the cleavage parameters are used is generating the canonical peptide pool. So technically we can let users generate different canonical peptide pools within the same index dir, because the annotation, genome, and proteome are all the same.

That's a good idea to run a dataset for sanity check. I have several PRs on the way so I'll just run the small collaborator's data (1 cell line) after all these PRs and before we release 1.3

**Comment by lydiayliu** [comment]
> The only place where the cleavage parameters are used is generating the canonical peptide pool. So technically we can let users generate different canonical peptide pools within the same index dir, because the annotation, genome, and proteome are all the same.

So with the current approach, the results would be wrong if there is a mismatch of cleavage parameters between `generateIndex` and `callVariant`? If so then I think updating that would be a good idea, we can even allow change in enzyme?

**Comment by zhuchcn** [comment]
With the current approach, callVariant will actually fail if the parameters don't match with that from `metadata.json`. Yeah we can change enzyme, too. I'll open an issue for this.

**Pull request merged**